include/crab/domains/region_domain.hpp

```diff
@@ -1325,6 +1428,48 @@ class region_domain final
     }
   }
 
+  /* MRU */
+  void commit_cache() {
+    if (!m_cache_dom.is_bottom()) {
+        // 1. join cache and base domain
+        m_base_dom |= m_cache_dom;
+        // 1.a update init
```

Comment:
I leave this part because I'm not sure whether we need to distinguish variable expanded from the base domain into cache domain.
For instance,
``` c
region_init(@V_1:region(int));
region_init(@V_2:region(int));

/* Cache: _|_, Sum: T */
s1 := make_ref(@V_1:region(int),8,as_0);
(@V_1:region(int),_0:ref) := gep_ref(@V_1:region(int),s1:ref);
/* Cache: _|_, Sum: T */
store_to_ref(@V_1:region(int),_0:ref,3:int32);
// cache missed. Because cache is empty, require populate cache (expand relationships).
// perform strong update
/* Cache: @V_1_copy = 3, Sum: T */
(@V_2:region(int),_1:ref) := gep_ref(@V_1:region(int),s1:ref + 4);
// Cache hit, perform strong update
store_to_ref(@V_2:region(int),_1:ref,5:int32);
/* Cache: {@V_1_copy = 3, @V_2_copy = 5}, Sum: T */
```
The question is whether we need a map to keep track of correspondence: `V_1_copy <-> V_1; V_2_copy <-> V_2` or we just have the same region variables both in cache domain and base domain?

Reply:
Just some context that will help @agurfinkel  to follow this but I'm not answering directly the question yet.

The base domain is the underlying numerical domain. So top-level abstract domains can express regions or arrays but ultimately everything is mapped to ghost (numerical) variables (we call it also base variables) in the base (numerical) domain.
Region variables are never mapped  directly to the base domain. This is simply, because a numerical domain doesn't know what to do with a region. So we need to map a region variable into actually one or more numerical variables. For instance, if the region contains an integer then we have a numerical variable representing that content of the region. If the region contains a reference then the region is mapped to two integers (one per offset and one per size of allocation site).  

So the expand operation (at least in the original region domain) happens at the base (numerical) domain level so there is no regions involved. The `init` and `count` environments map region variables to some piece of information (whether the region is initialized and how many references contains, respectively). The number of region variables is always fixed in the CrabIR program so the abstract domain never creates region variables at runtime.